### PR TITLE
move make_array array_append array_prepend array_concat function to datafusion-functions-array crate

### LINKED ISSUE #9322: move make_array array_append array_prepend array_concat function to datafusion-functions-array crate
### Is your feature request related to a problem or challenge?

sub-issue of #9285

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### PR DESCRIPTION
**Pull request by guojidan**
## Which issue does this PR close?

Closes #9322.

## What changes are included in this PR?

move make_array array_append array_prepend array_concat function to datafusion-functions-array crate

### CODE REVIEW

**Comment by guojidan** [comment]
because #9343 pr lasting for too long，difficult to rebase or merge, so I open a new pr

**Review by jayzhan211 — APPROVED** [review]
I take a quick glance and I think it is ready to merge.

There are 2 things that I'm trying to figure out to improve further
1. Make sure we have `Scalar` that we need `make_scalar_function_with_hints`, since I think we should only get `Array` for array functions, if we really have `Scalar`, it seems something wrong to me.
2. Simplify cfg feature flag, I would like to change `OperatorToFunction` to `OperatorToArrayFunciton` and see whether we can import the entire rule (file) optionally.


They don't need to be solved in this PR, but I might take an improvement on these, if I figure out it is possible.

File: datafusion/optimizer/Cargo.toml [hunk 1]
```diff
@@ -44,6 +45,7 @@ async-trait = { workspace = true }
 chrono = { workspace = true }
 datafusion-common = { workspace = true, default-features = true }
 datafusion-expr = { workspace = true }
+datafusion-functions-array = { workspace = true, optional = true }
```
Comment (jayzhan211) [inline datafusion/optimizer/Cargo.toml:48]: @alamb I think having optional importing is fine. Is there any other downside of this?

**Comment by jayzhan211** [comment]
let me rebase and fix the conflicts

**Comment by jayzhan211** [comment]
> 1. Make sure we have `Scalar` that we need `make_scalar_function_with_hints`, since I think we should only get `Array` for array functions, if we really have `Scalar`, it seems something wrong to me.

I think we need `make_scalar_function_with_hints`, Scalar::List are represented in ColumnValue::Scalar, and others are represented in ColumnValue::Array. Although we don't do the optimized Scalar path for array functions, to be consistent with the definition of ColumnValue, it may be nice be convert it to ColumnValue::Scalar if we got Scalar::List.

File: datafusion/functions-array/src/udf.rs [hunk 2]
```diff
@@ -392,3 +397,259 @@ impl ScalarUDFImpl for ArrayNdims {
         &self.aliases
     }
 }
+
+make_udf_function!(
+    ArrayAppend,
+    array_append,
+    array element,                                         // arg name
+    "appends an element to the end of an array.", // doc
+    array_append_udf                              // internal function name
```
Comment (Weijun-H) [inline datafusion/functions-array/src/udf.rs:406]:
```suggestion
    array element,
    "appends an element to the end of an array.",
    array_append_udf
```

File: datafusion/functions-array/src/udf.rs [hunk 3]
```diff
@@ -392,3 +397,259 @@ impl ScalarUDFImpl for ArrayNdims {
...
+        &self.aliases
+    }
+}
+
+make_udf_function!(
+    ArrayPrepend,
+    array_prepend,
+    element array,                                                // arg name
+    "Prepends an element to the beginning of an array.", // doc
+    array_prepend_udf                                    // internal function name
```
Comment (Weijun-H) [inline datafusion/functions-array/src/udf.rs:460]:
```suggestion
    element array,
    "Prepends an element to the beginning of an array.",
    array_prepend_udf
```

File: datafusion/functions-array/src/udf.rs [hunk 4]
```diff
@@ -392,3 +397,259 @@ impl ScalarUDFImpl for ArrayNdims {
...
+    fn aliases(&self) -> &[String] {
+        &self.aliases
+    }
+}
+
+make_udf_function!(
+    ArrayConcat,
+    array_concat,
+    "Concatenates arrays.", // doc
+    array_concat_udf        // internal function name
```
Comment (Weijun-H) [inline datafusion/functions-array/src/udf.rs:513]:
```suggestion
    "Concatenates arrays.",
    array_concat_udf
```

File: datafusion/functions-array/src/udf.rs [hunk 5]
```diff
@@ -392,3 +397,259 @@ impl ScalarUDFImpl for ArrayNdims {
...
+    fn aliases(&self) -> &[String] {
+        &self.aliases
+    }
+}
+
+make_udf_function!(
+    MakeArray,
+    make_array,
+    "Returns an Arrow array using the specified input expressions.", // doc
+    make_array_udf // internal function name
```
Comment (Weijun-H) [inline datafusion/functions-array/src/udf.rs:591]:
```suggestion
    "Returns an Arrow array using the specified input expressions.",
    make_array_udf
```

**Review by Weijun-H — COMMENTED** [review]
LGTM. thanks @guojidan

**Review by Weijun-H — APPROVED** [review]

File: datafusion/functions-array/src/udf.rs (hunk 2 above)
Comment (Weijun-H) [reply datafusion/functions-array/src/udf.rs:406]: I believe that we can get rid of these comments since the structure is apparent.

**Comment by jayzhan211** [comment]
@Weijun-H I think we can merge this and improve others on follow up PR

**Comment by jayzhan211** [comment]
but I have no idea why the test is running endless

**Comment by Weijun-H** [comment]
> but I have no idea why the test is running endless

It should be an issue because it happened locally.

**Comment by jayzhan211** [comment]
I want to split them to different files to avoid crazy conflict

**Comment by jayzhan211** [comment]
It seems super slow to run `array.slt` with this PR

**Comment by jayzhan211** [comment]
@guojidan Besides fixing conflicts. I split those functions to their own files, and rename to `make_scalar_funciton` (there should be a better name).

let me merge this PR first, and file another PR for any other changes.

Thanks @guojidan and @Weijun-H

**Pull request merged**